core/src/main/java/hudson/ProxyConfiguration.java

```diff
@@ -259,6 +259,40 @@ public PasswordAuthentication getPasswordAuthentication() {
         return con;
     }
 
+    /** This method invokes {@link ProxyConfiguration#open(URL)} setting {@link HttpURLConnection#setInstanceFollowRedirects(boolean)} to true if the opened connection is http.
+     * In case of a redirection it will be follow it, <b>even if the redirection changes the protocol (from http to https or vice versa) using the value in Location header</b>.
+     * @since TODO
+     */
+    public static URLConnection openURLAndRedirect(URL url) throws IOException {
+        URLConnection con = ProxyConfiguration.open(url);
+        // A redirection from http to https (or vise versa) returns a 302 response status. Force redirection
+        int responseCode = HttpURLConnection.HTTP_OK;
+        // TODO responseCode always will be OK for non-http connections. Processing needed
+        if (con instanceof HttpURLConnection) {
+            // prevent problems from misbehaving plugins disabling redirects by default
+            HttpURLConnection httpCon = ((HttpURLConnection) con);
+            httpCon.setInstanceFollowRedirects(true);
+            responseCode = httpCon.getResponseCode();
+            if (HttpURLConnection.HTTP_OK != responseCode && HttpURLConnection.HTTP_MOVED_TEMP != responseCode && HttpURLConnection.HTTP_MOVED_PERM != responseCode && HttpURLConnection.HTTP_SEE_OTHER != responseCode) {
+                // No OK Status and no redirection
+                throw new IOException("Could not open " + url + ". Response code: " + responseCode);
+            } else if (HttpURLConnection.HTTP_OK != responseCode) {
+                // In case of redirection, we have to connect to the new URL
+                String redirection = con.getHeaderField("Location");
+                con = ProxyConfiguration.open(new URL(redirection));
```

Comment:
It is unclear to me how this could happen when we're telling the connection to follow redirects. Isn't this the whole point of that flag that we don't need to do this manually?

And if we're doing it manually, why is this line not a recursive call?

Reply:
Never mind, this implements the cross protocol redirection which `setInstanceFollowRedirects` doesn't.

That still doesn't explain why this isn't a recursive call though.

Reply:
@daniel-beck It is a java issue. You can see the details in this [link].(https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4620571)

By the way, I will turn the call into a recursive one.